### ISSUE TITLE
torch.nn.LayerNorm support for arbitrary axis in order to allow NCHW application

### ISSUE THREAD

**Issue by dzhulgakov** [pytorch/pytorch#71465]
### 🚀 The feature, motivation and pitch

LayerNorm starts to be applied to image data on per-channel basis (e.g. in ConvNeXt model).

`torch.nn.LayerNorm` support normalization only on the last several dimensions. If the input is NCHW tensor (default layout) it requires explicit NHWC conversion in order to do normalization on channels only. Example from ConvNeXt implementation: https://github.com/facebookresearch/ConvNeXt/blob/main/models/convnext.py#L15

`LayerNorm` could instead allow to specify which axis the normalization should be performed over and be backed by an efficient kernel. Example in Keras receiving `axis` argument: https://www.tensorflow.org/api_docs/python/tf/keras/layers/LayerNormalization 

Additionally, it'd be nice to also support `memory_format=channels_last` setting for the input tensor. Though it's an orthogonal point - this issue talks about semantics, not about the underlying memory format.

P.S. related discussion: https://twitter.com/wightmanr/status/1481383606782087168

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/51455

**Comment by albanD** [comment]
Note that for no batch dim support, we most likely don't want to add any case that would make the existence or not of a batch dim ambiguous.

**Comment by rwightman** [comment]
@DmitryUlyanov Can we bump this in priority? More and more networks would benefit from this. The lack of a native impl appears to be governing some design decisions in cnn-transformer hybrid models implemented in PyTorch (or causing issues with Tensorflow/JAX originating models where this isn't a concern).

It's not only ConvNeXt now, but EdgeNeXt, CoaTNet (Google version), NesT, VisFormer, PoolFormer (would have used it but opted for GroupNorm(groups=1) instead, not equivalent). Now MobileVit2 appears to be trying to use it but opted for GroupNorm(groups=1), possibly without realizing the lack of equivalence based on how they named it (https://github.com/apple/ml-cvnets/issues/34)

**Comment by DmitryUlyanov** [comment]
I am not sure why you tagged me here, but why won't you just use [InstanceNorm](https://pytorch.org/docs/stable/generated/torch.nn.InstanceNorm2d.html)?

**Comment by rwightman** [comment]
> I am not sure why you tagged me here, but why won't you just use [InstanceNorm](https://pytorch.org/docs/stable/generated/torch.nn.InstanceNorm2d.html)?

Sorry, I was a bit quick on the @ autocomplete there :/

**Comment by jbschlosser** [comment]
Making high-pri due to activity.

**Comment by vadimkantorov** [comment]
> To easiest way to get what we want in PyTorch right now is `F.layer_norm(x.permute(0, 2, 3, 1), self.normalized_shape, self.weight, self.bias, self.eps).permute(0, 3, 1, 2)` ... however those permutes aren't free, and this mucks performance up when you're trying to use memory_format=channels_last.

@ngimel also highlighted that channel_last isn't generally well supported by Norm's: https://github.com/pytorch/pytorch/issues/72341 :(

**Comment by rwightman** [comment]
@vadimkantorov yeah, I'd really want to have channels_last support, it would be much less useful without.

**Comment by ngimel** [comment]
Channels-last layer-norm can kinda be done today, with permute (free, only changes sizes and strides), regular layer norm and unpermute (again, free, and will nicely result in channels-last tensor to send to the next conv).

**Comment by rwightman** [comment]
@ngimel as in this ? `F.layer_norm(x.permute(0, 2, 3, 1), self.normalized_shape, self.weight, self.bias, self.eps).permute(0, 3, 1, 2)`?

I've spent some time running through iterations of this, and while the above is 'okay' there didn't appear to be any improvement using CL vs not, where as if you replaced those LN with BN there was, I was able to boost throughput noteably w/ channels_last doing some  hacks (checking for contiguous, and switching between that impl and a full custom mean/var as you see in convnext) with ~10+% bump on Ampere cards, but killed throughput on older cards.

**Comment by ngimel** [comment]
@rwightman what would be the command to run convnext channels-last fp16 benchmark?

**Comment by ezyang** [comment]
@ngimel If layer norm kernel already has the smarts for this we should just add the memory format support, should be easy right?

**Comment by ngimel** [comment]
layer norm kernel doesn't have the smarts, it works on contiguous tensors and normalizes on the last dimension.

**Comment by ezyang** [comment]
Do you think our compiler like nvFuser is good enough to produce a good kernel? Maybe we should just have someone sit down and write the Triton kernel

**Comment by ngimel** [comment]
see https://github.com/rwightman/pytorch-image-models/pull/1340

**Comment by xvjiarui** [comment]
Just out of curiosity, is the channel last LayerNorm or [LayerNorm2d from timm](https://github.com/rwightman/pytorch-image-models/blob/d7b55a9429f3d56a991e604cbc2e9fdf1901612f/timm/models/layers/norm.py#L26) equivalent to Positional Norm here https://github.com/Boyiliee/Positional-Normalization?

**Comment by hgaiser** [comment]
Is there any progress on this? I'm running into the same issue when implementing ViTDet in torchvision.

**Comment by pbelcak** [comment]
Bumpidy bump. Would also greatly benefit from this being implemented (usecase CV models with NCHW tensor layout).

Looking at [the support for fused `index_select` with matmul](https://github.com/pytorch-labs/gpt-fast/pull/57), maybe instead of having a completely new operator, we could just fuse `.transpose(k, -1)` (or, more generally, `Tensor.permute`) with layer norm?

**Comment by ezyang** [comment]
Someone should check if PT2 layer norm does well on this

**Comment by elisim** [comment]
Any updates on this? Newer architectures could greatly benefit from it.

@ezyang, if I understood you correctly, PT2 hasn't implemented the axis API yet.

**Comment by ezyang** [comment]
Yes, but if you feed in an appropriately transposed tensor (so NHWC) we might be able to avoid the contiguous call

**Comment by elisim** [comment]
@ezyang, to clarify, are you saying that for x with shape NCHW

```python
F.layer_norm(x.permute(0, 2, 3, 1), self.normalized_shape, self.weight, self.bias, self.eps).permute(0, 3, 1, 2)
```

might perform fast in PT2 even if the tensor isn’t contiguous?

**Comment by ezyang** [comment]
yup

**Comment by Siddhant231xyz** [comment]
I opened https://github.com/pytorch/pytorch/pull/196926 adding a `dim` argument to `nn.LayerNorm` / `F.layer_norm`, so the ConvNeXt-style channels-only case becomes `nn.LayerNorm(C, dim=1)` on an NCHW tensor. It is the movedim -> layer_norm -> movedim formulation discussed above (views only, channels_last preserved, fuses under torch.compile) with the default behaviour untouched. AI disclosure: drafted with an AI assistant under my direction and reviewed/tested by me. Happy to follow up with a fused channels_last kernel if wanted.
